Specification use linkml

Silvanoc opened the pull request:
### Description

Use [LinkML](https://linkml.io) to specify the resources needed in the [Desired State API](https://specification.margo.org/margo-api-reference/workload-api/desired-state-api/desired-state/) and [Application Package Definition API](https://specification.margo.org/app-interoperability/application-package-definition/).

The use of LinkML provides following advantages over the current workflow:
1. Formal specification for the resources. Including tooling to validate the consistency of the specification.
2. Tooling to validate resources conformance with the resource specification, including constraints.
3. Tooling to automatically generate the documentation (MarkDown that gets converted into HTML with mkdocs).
4. Tooling to generate JSON-Schema that can be integrated into other tools (e.g. SDKs).

#### Issues Addressed

List and link all the issues addressed by this PR.

#### Change Type

Please select the relevant options:

- [ ] Fix (change that resolves an issue)
- [x] New enhancement (change that adds specification content)
- [ ] Content edits (change that edits existing content)

#### Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) document.
- [x] My changes adhere to the established patterns, and best practices.

Silvanoc committed:
feat(desired-state): add linkml specification

Add LinkML specification of the DesiredState resource used in the
desired-state-api.

This patch also adds GitHub Actions to validate the specification with
(counter-)examples and to generate the needed MarkDown documents to
generate the HTML documentation.

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

arne-broering committed:
feat(app-definition): add LinkML spec

Signed-off-by: Arne Broering <arne.broering@siemens.com>
Co-authored-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
chore(utils): fix doc validation and generation

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
feat(app-definition): minor changes

Explicitly declare inlined lists

Fix representation of attributes in docu

Improve representation of arrays of references

Explicitly declare inlined arrays

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
chore(git): minor changes

Update .gitignore

Remove to-be-generated markdown document

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
chore(tree): improve tree structure

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
docs(desired-state): improve counter-examples

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

phil-abb commented:
@Silvanoc What do I need to do to get these two pages that were moved to LinkML to render locally? I've pulled down the repo and when I use `mkdocs server` to show the rendered documentation locally these two pages are not there anymore.

![image](https://github.com/user-attachments/assets/dc7f1917-35b8-479d-a253-a7acd26aff90)

phil-abb commented on src/app-interoperability/resources/class.md.jinja2:
This file is empty. Should it be removed?

Silvanoc replied:
No, it's a needed empty placeholder.

Silvanoc commented:
> @Silvanoc What do I need to do to get these two pages that were moved to LinkML to render locally? I've pulled down the repo and when I use `mkdocs server` to show the rendered documentation locally these two pages are not there anymore.

@phil-abb have you executed [`doc-generation/generate-documentation.bash`](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R99)? I forgot to add that information to the documentation.

phil-abb commented:
> Have you executed [`doc-generation/generate-documentation.bash`](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R99)? I forgot to add that information to the documentation.

I haven't done anything because I'm not sure what I need to do. I can try running this script. 

We should document what needs to be done. @ajcraig should this be added to the CONTRIBUTING.md, or somewhere else? We have a version of the CONTRIBUTING.md somewhere with information on how to run stuff locally but it doesn't look like that was ever merged into pre-draft.

phil-abb commented:
> Have you executed [`doc-generation/generate-documentation.bash`](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R99)? 

@Silvanoc  I tried running it but got this result:

```bash
vscode ➜ ~/docs (specification_use-linkml) $ ./doc-generation/generate-documentation.bash 
/home/vscode/docs/src/app-interoperability
Creating virtualenv specification-lHOaJu9a-py3.12 in /home/vscode/.cache/pypoetry/virtualenvs
None
/home/vscode/docs/src/margo-api-reference/workload-api/desired-state-api
None
Error: Config file 'mkdocs.yml' does not exist.
vscode ➜ ~/docs (specification_use-linkml) $ ls
CONTRIBUTING.md  LICENSE  README.md  SECURITY.md  assets  certs  doc-generation  mkdocs.yml  poetry.lock  pyproject.toml  src  system-design
vscode ➜ ~/docs (specification_use-linkml) $ 
```

phil-abb commented on .github/workflows/pages.yml:
```diff
@@ -45,7 +140,13 @@ jobs:
+      - name: Get generated documentation from previous jobs
+        uses: actions/download-artifact@v4
+        with:
+          name: desired-state.md
```
This is a duplicate of what is above. Is this intentional? Is there supposed to be something here for the application package definition file?

Silvanoc replied:
Do you mean the [`upload-artifact` step](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R101) with "above"? If that's so, then it's not really duplicate. The above one "uploads" the artifact to make it usable for the next step that "downloads" that artifact.

Silvanoc replied:
@phil-abb please resolve this conversation yourself if satisfied with the explanation. Otherwise I would need more information to better understand what you mean. Thanks.

phil-abb replied:
@Silvanoc ah, yeah, I missed that. What I'm still confused about is there is nothing in here for the application package definition. Should there be something in here for that?

Silvanoc replied:
Very good point! @arne-broering took care of the application package definition and probably forgot to add that part. I suppose that he's been testing it locally and therefore has never realized that it is missing.

phil-abb commented on .github/workflows/pages.yml:
```diff
@@ -3,16 +3,111 @@
+  push:
+    branches:
+      - pre-draft
+      - specification_use-linkml
```
Should we remove references to the "specification_use-linkml" branch from this file before merging?

Silvanoc replied:
As mentioned in the "IMPORTANT NOTE" in the description of this PR, there are at least three commits (the message header starts with `tmp:`) only needed to review the branch that should be removed before merging. Therefore I put the PR as draft.

Here you can see at least two of the mentioned commits:
![Screenshot 2025-01-24 at 17 42 56](https://github.com/user-attachments/assets/34d9d393-d167-4cf8-b574-986fa90580bd)

This is one of the this that disappears with the removed commits.

phil-abb commented on .github/workflows/pages.yml:
```diff
@@ -45,7 +140,13 @@ jobs:
       - name: GitHub Pages Deploy
         # only deploy if we are on the default branch
-        if: ${{ github.ref == 'refs/heads/pre-draft' }}
+        if: ${{ github.ref == 'refs/heads/pre-draft' || github.ref == 'refs/heads/specification_use-linkml' }}
```
Remove branch reference here as well?

Silvanoc replied:
As mentioned in the "IMPORTANT NOTE" in the description of this PR, there are at least three commits (the message header starts with `tmp:`) only needed to review the branch that should be removed before merging. Therefore I put the PR as draft.

Here you can see at least two of the mentioned commits:
![Screenshot 2025-01-24 at 17 42 56](https://github.com/user-attachments/assets/34d9d393-d167-4cf8-b574-986fa90580bd)

This is one of the this that disappears with the removed commits.

phil-abb commented on system-design/margo-api-reference/workload-api/desired-state-api/.gittree:
This is an empty file. Should it be included?

Silvanoc replied:
Having it like this `git clone` creates the directory containing the file. Otherwise we would need extra `mkdir`s to create the directory.

Silvanoc replied:
@phil-abb please resolve this conversation yourself if satisfied with the current status. Otherwise please let me know what you would prefer and I can adapt it. Thanks.

phil-abb commented on system-design/CNAME.temporarily.disabled:
The should be reverted before the merge

Silvanoc replied:
As mentioned in the "IMPORTANT NOTE" in the description of this PR, there are at least three commits (the message header starts with `tmp:`) only needed to review the branch that should be removed before merging. Therefore I put the PR as draft.

Here you can see at least two of the mentioned commits:
![Screenshot 2025-01-24 at 17 42 56](https://github.com/user-attachments/assets/34d9d393-d167-4cf8-b574-986fa90580bd)

This is one of the this that disappears with the removed commits.

phil-abb reviewed (commented):
Here are some initial comments. I'll probably have some more once I can get it to render locally.

Silvanoc commented:
> > Have you executed [`doc-generation/generate-documentation.bash`](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R99)?
> 
> @Silvanoc I tried running it but got this result:

@phil-abb have you prepared the development environment like in the GitHub Action?

You will need Python Poetry ([the GitHub Action uses pipx for this purpose](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R86)).

[Installing the dependencies like in the GitHub Action](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R95) should ensure that you have installed everything you need: `poetry install --all-extras`

All that information belongs into a `How to test locally` or similar, but until we have it, I will help you.

phil-abb commented:
> @ajcraig should this be added to the CONTRIBUTING.md, or somewhere else? We have a version of the CONTRIBUTING.md somewhere with information on how to run stuff locally but it doesn't look like that was ever merged into pre-draft.

@ajcraig / @Silvanoc - This is [the branch](https://github.com/margo/specification/blob/ajc/mkdocs-tutorial/CONTRIBUTING.md) that had those changes. @ajcraig Was there a reason we never merged this in, or did we just forget about it?

phil-abb commented:
> [Installing the dependencies like in the GitHub Action](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R95) should ensure that you have installed everything you need: `poetry install --all-extras`
> 

@Silvanoc This is what I get when I run that command. I stripped out several `- installing XYZ` lines to keep it short

```bash
vscode ➜ ~/docs (specification_use-linkml) $ poetry install --all-extras
Creating virtualenv specification-lHOaJu9a-py3.12 in /home/vscode/.cache/pypoetry/virtualenvs
Installing dependencies from lock file

Package operations: 97 installs, 0 updates, 0 removals

  - Installing typing-extensions (4.12.2)
  - Installing annotated-types (0.7.0)
  ...
  - Installing mkdocs-markdownextradata-plugin (0.2.6)
  - Installing mkdocs-material (9.5.46)
  - Installing tox (4.23.2)

Installing the current project: specification (0.1.0)
Error: The current project could not be installed: No file/folder found for package specification
If you do not want to install the current project use --no-root.
If you want to use Poetry only for dependency management but not for packaging, you can disable package mode by setting package-mode = false in your pyproject.toml file.
If you did intend to install the current project, you may need to set `packages` in your pyproject.toml file.
```

Silvanoc commented:
> > @ajcraig should this be added to the CONTRIBUTING.md, or somewhere else? We have a version of the CONTRIBUTING.md somewhere with information on how to run stuff locally but it doesn't look like that was ever merged into pre-draft.
> 
> @ajcraig / @Silvanoc - This is [the branch](https://github.com/margo/specification/blob/ajc/mkdocs-tutorial/CONTRIBUTING.md) that had those changes. @ajcraig Was there a reason we never merged this in, or did we just forget about it?

Without knowing if the proposal would be accepted I didn't invest much time on ensuring that everything is well documented. If LinkML is being accepted for the specification of resources for the API, then I would need to improve the documentation and probably ensure that is can be easily run using Dev Containers.

Silvanoc committed:
chore(python): fix project configuration

Poetry needs the project configuration to clearly state if the
repository is a Python package or not (in which case Poetry is only used
to install dependencies and tools). This commit fixes it.

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc commented:
> > [Installing the dependencies like in the GitHub Action](https://github.com/margo/specification/pull/62/files#diff-267452bb53d388a7b14551d9fc53a797edc124dc9085ecc79f5339277a1e53b5R95) should ensure that you have installed everything you need: `poetry install --all-extras`
> 
> @Silvanoc This is what I get when I run that command. I stripped out several `- installing XYZ` lines to keep it short

If I am right, you are using Poetry 2 and they have made an error where before it was only a warning. I have just pushed a new commit adding the (hopefully) needed change to get it running.

phil-abb commented:
> If I am right, you are using Poetry 2 and they have made an error where before it was only a warning. I have just pushed a new commit adding the (hopefully) needed change to get it running.

@Silvanoc That fixed the error I was getting with the install command. I'm still getting the `Error: Config file 'mkdocs.yaml' does not exist` message though.

I am trying to run this from the dev container. I made the following changes to the ./devcontainer/dockerfile to add Poetry and LinkML and updated to use python 3.12

```yaml
FROM mcr.microsoft.com/vscode/devcontainers/python:3.12-bookworm

RUN pip install mkdocs \
    && pip install mkdocs-markdownextradata-plugin \
    && pip install mkdocs-material \
    && pip install poetry \
    && pip install linkml 
```

Silvanoc commented:
> > If I am right, you are using Poetry 2 and they have made an error where before it was only a warning. I have just pushed a new commit adding the (hopefully) needed change to get it running.
> 
> @Silvanoc That fixed the error I was getting with the install command. I'm still getting the `Error: Config file 'mkdocs.yaml' does not exist` message though.

That's weird.
I do not use dev containers, therefore I cannot tell if it has something to do with it or not. In my machine, the one from Arne and GitHub Actions it works. And the file is there!

> I am trying to run this from the dev container. I made the following changes to the ./devcontainer/dockerfile to add Poetry and LinkML and updated to use python 3.12
> 
> ```yaml
> FROM mcr.microsoft.com/vscode/devcontainers/python:3.12-bookworm
> 
> RUN pip install mkdocs \
>     && pip install mkdocs-markdownextradata-plugin \
>     && pip install mkdocs-material \
>     && pip install poetry \
>     && pip install linkml 
> ```

I will see if I can get dev containers up and running and try to reproduce your issue.

Silvanoc commented:
@phil-abb could you please push your additions/modifications to the dev container configuration on top of my branch? That way I can try to reproduce it on my side.

Silvanoc commented:
> > If I am right, you are using Poetry 2 and they have made an error where before it was only a warning. I have just pushed a new commit adding the (hopefully) needed change to get it running.
> 
> @Silvanoc That fixed the error I was getting with the install command. I'm still getting the `Error: Config file 'mkdocs.yaml' does not exist` message though.
> 
> I am trying to run this from the dev container. I made the following changes to the ./devcontainer/dockerfile to add Poetry and LinkML and updated to use python 3.12
> 
> ```yaml
> FROM mcr.microsoft.com/vscode/devcontainers/python:3.12-bookworm
> 
> RUN pip install mkdocs \
>     && pip install mkdocs-markdownextradata-plugin \
>     && pip install mkdocs-material \
>     && pip install poetry \
>     && pip install linkml 
> ```

If all the dependencies are manually installed, then Poetry is not needed (the line `&& pip install poetry` would not be needed). But I would rather install with Poetry, because it freezes the versions of the dependencies, making the whole setup much more reproducible.

If using Poetry, then only that dependency is needed, but then the dependencies need to be installed using Poetry with an additional `RUN` step.

phil-abb committed:
Updated dev container .dockerfile

Signed-off-by: Philip <philip.presson@us.abb.com>

phil-abb commented:
> @phil-abb could you please push your additions/modifications to the dev container configuration on top of my branch? That way I can try to reproduce it on my side.

@Silvanoc Updates pushed. I tried just installing poetry in the .dockerfile but even after running the `poetry install --all extras` it didn't recognize mkdocs was installed. I'm sure I'm doing something wrong, I haven't had much experience with Python.

Silvanoc commented:
> > @phil-abb could you please push your additions/modifications to the dev container configuration on top of my branch? That way I can try to reproduce it on my side.
> 
> @Silvanoc Updates pushed. I tried just installing poetry in the .dockerfile but even after running the `poetry install --all extras` it didn't recognize mkdocs was installed. I'm sure I'm doing something wrong, I haven't had much experience with Python.

The dev container setup makes it all more difficult to diagnose than the pure python approach. But I suppose that you (as well as many other Margo participants) are running Windows and therefore need them.

I will have to leave it as it is for now. But I will try to diagnose it ASAP.

dev containers are, unfortunately, very opinionated. They probably run smoothly on Windows. In theory they can run on MacOS, but are somehow requiring docker desktop (which I do not use to run containers).

phil-abb commented:
> I will have to leave it as it is for now. But I will try to diagnose it ASAP.

@Silvanoc I was able to get things to work. I'm still getting the error when I run the `./doc-generation/generate-documentation.bash` script but it still creates the two documents.

Also, when just installing Poetry in the .dockerfile I was able to get mkdocs to run by calling `poetry run mkdocs serve --dev-adder 0.0.0.0:8000` I didn't realize I needed the `poetry run` part when doing it this way.

phil-abb committed:
Updated dev container .dockerfile to just install poetry

Signed-off-by: Philip <philip.presson@us.abb.com>

phil-abb commented on src/app-interoperability/resources/index.md.jinja2:
I'm not sure if this is because this is still a work in progress, or if it was accidentally missed, but there is content missing from this page.

The text in the "Defining configurable application parameters" is missing what is in the [original text](https://specification.margo.org/app-interoperability/application-package-definition/)

Also the "validation rules" are only partially there but it's not right in the sense it's mixed in with the schema attributes and it only covers the text rules. Is this because LinkML doesn't handle what we need to do here for the different validation rules based on the schema type?

Silvanoc replied:
@arne-broering is working on that part, which is WIP. I recommended him not to go "all in" until we know if LinkML is being accepted, so it is on-purpose that is not ready yet 🙂

Silvanoc replied:
@arne-broering has fixed some issues in the "application package definition" with [this PR](https://github.com/Silvanoc/margo-specification/pull/2), @phil-abb I suppose that now it looks closer to your expectation.

arne-broering replied:
@Silvanoc thanks for merging my changes here. Indeed, I reflected the validation rules now as subclasses and adjusted the template so that all of them are being displayed now.

arne-broering committed:
feat(app-definition): jinja2 template adjustments

Signed-off-by: Arne Broering <arne.broering@siemens.com>

Silvanoc commented:
> > I will have to leave it as it is for now. But I will try to diagnose it ASAP.
> 
> @Silvanoc I was able to get things to work. I'm still getting the error when I run the `./doc-generation/generate-documentation.bash` script but it still creates the two documents.

I suppose that you mean this error: `Error: Config file 'mkdocs.yaml' does not exist`. I wonder how the generation works with that configuration missing.

Since our setups (mine is MacOS + vim + Poetry) are so different I cannot reproduce that issue on my side. But I am trying to get dev containers up and running on my system to be closer to each other. I wonder if the dev containers approach is not too tool-opinionated for such an open project. Would it work with Linux? Would it work without VSCode?

> Also, when just installing Poetry in the .dockerfile I was able to get mkdocs to run by calling `poetry run mkdocs serve --dev-adder 0.0.0.0:8000` I didn't realize I needed the `poetry run` part when doing it this way.

Let me correct myself. `poetry run` is only needed when you don't want to install all the tools natively. In fact it is possible to freeze the versions without Poetry using a [`requirements.txt` file](https://www.freecodecamp.org/news/python-requirementstxt-explained/) exported from the Poetry version freeze. For a container that is probably the right way to approach it instead of installing Poetry.

Silvanoc committed:
chore(python): simplify dependency management

Requiring Poetry>2 we can use the same dependency definition method for
dependency resolution with PIP and Poetry.

This also simplifies the creation of the dev-container.

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>
Suggested-by: Michael Adler <michael.adler@siemens.com>

Silvanoc committed:
fix(chore): wrong path in document generation

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc commented:
> I suppose that you mean this error: Error: Config file 'mkdocs.yaml' does not exist. I wonder how the generation works with that configuration missing.

That error was the result of a bug in the shell scrip that I have resolved in the last commit.

> Let me correct myself. `poetry run` is only needed when you don't want to install all the tools natively. In fact it is possible to freeze the versions without Poetry using a [`requirements.txt` file](https://www.freecodecamp.org/news/python-requirementstxt-explained/) exported from the Poetry version freeze. For a container that is probably the right way to approach it instead of installing Poetry.

I have adapted the Dockerfile of the dev container to use a multi-stage build that results in the installation of dependencies in the expected versions with pip. That is much more intuitive for someone not so deep into Python and Poetry.

@phil-abb with this change you don't need to use `poetry run`, since all dependencies are globally installed in the container image. Now you should be able to generate without the error message.

phil-abb commented:
> @phil-abb with this change you don't need to use...

@Silvanoc Great. After I return home from the workshop.

arne-broering committed:
making sure generated html looks as before

Signed-off-by: Arne Broering <arne.broering@siemens.com>

Silvanoc committed:
chore(utils): adapt shell scripts for flexibility

Make document generation scripts more flexible so that they work on
systems using Poetry and on systems usind dev containers.

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc commented:
Scripts adapted to be run on:
- systems providing Python Poetry natively for contained installation of dependencies, as well as on
- systems using dev-containers for contained installation of dependencies

Both of them tested on my system seem to work as expected.

arne-broering committed:
changed validationrule classes to become subclasses of schema

Signed-off-by: Arne Broering <arne.broering@siemens.com>

phil-abb commented:
I'm perfectly fine with what I'm seeing here and the use of LinkML.

@Silvanoc What do you feel needs to happen for this PR to be moved out of the `draft` state?

cc @arne-broering / @ajcraig

michaeladler commented on pyproject.toml:
```diff
@@ -10,13 +10,17 @@ description = "Margo Specification"
 authors = []
 license = "Open Web Foundation - CLA Copyright Grant 0.9"
 readme = "README.md"
+package-mode = false
```
Following PEP-518 the file should now rather look like that:

```toml
[project]
name = "specification"
version = "0.1.0"
description = "Margo Specification"
authors = []
license = "Open Web Foundation - CLA Copyright Grant 0.9"
readme = "README.md"
requires-python = ">=3.12,<4.0"
dependencies = [
  'mkdocs>=1.6.1',
  'mkdocs-markdownextradata-plugin>=0.2.6',
  'mkdocs-material>=9.5.44',
  'linkml>=1.8.5',
]

[tool.poetry]
requires-poetry = ">=2.0"
package-mode = false

[tool.poetry.group.dev.dependencies]
tox = "^4.23.2"

[build-system]
requires = ["poetry-core>=2.0"]
build-backend = "poetry.core.masonry.api"
```

michaeladler commented on .devcontainer/.dockerfile:
```diff
@@ -1,7 +1,22 @@
+
+COPY pyproject.toml poetry.lock ./
+
+RUN poetry export --format=requirements.txt > requirements.txt
```
No need for this anymore, you can just do `pip install pyproject.toml` with the `pyproject.toml` I suggested below

Silvanoc committed:
fix: broken link

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
chore(python): update poetry.lock

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
fix(ci): get all generated documents

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
chore: add pre-push git hook to validate examples

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
fix(app-definition): schema

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc committed:
docs(general): how to generate the documentation

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc commented:
@phil-abb and @arne-broering, I have cleaned-up those commits needed to deploy the documentation on my fork and I have added some information on [how to document the resources using LinkML](https://github.com/margo/specification/pull/62/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R81). I am triggering now a new review from your side.

Silvanoc commented:
BTW, @arne-broering , @phil-abb you can see the result of the last deployment on my fork before clean-up [here](https://silvanoc.github.io/margo-specification/). The clean-up has only removed the commits needed for temporary deployment on my fork and beautified the Git history a bit (without changes in the final code).

phil-abb commented on system-design/orchestration/workload/workload-orchestration-edge-onboarding.md:
````diff
@@ -13,7 +13,56 @@ In order for the workload orchestration solution to manage the edge device's wor
 
-![Margo Management Interface Operational Flow Diagram (svg)](../../figures/margo-interface-generic.drawio.svg)
+``` mermaid
+sequenceDiagram
````
I have concerns about this sequence diagram being in this part of the specification because this is getting heavily into implementation details and outside the scope of the specification. For example, whether or not there is a Workload Fleet Manager Frontend is nothing the specification should be indicating because it's a design choice for the WFM supplier to make. With Margo, we want to have as many black boxes as possible and I think this diagram is leaking implementation details from the black box.

I see value in having these types of sequence diagrams though for us to understand what some workflows might be, I just don't know where they should go. Either we need to think about splitting this "specification" document up differently so we have sections dedicated to the technical details focusing only on the specification and sections talking about vision, use cases, example workflows, etc. Or we put these types of things in a separate document/repo to help the community be on the same page with how we think things might be used.

Silvanoc replied:
@phil-abb I have moved these kind of additions to a separate branch. Therefore resolving this conversation.

phil-abb commented on src/app-interoperability/resources/index.md.jinja2:
```diff
@@ -0,0 +1,201 @@
+
+This section defines the application package provided by an “Application Developer” who has implemented the application and aims to provide it to Margo-conformant systems. The application package comprises:
+
+- The **application description**: a YAML document (type = `ApplicationDescription`), which is stored in a file (e.g., `margo.yaml`) and contains information about the application's [metadata](#metadata-attributes) (e.g., description, icon, release notes, license file, etc.), application supported [deployment configurations](#deploymentprofile-attributes) (e.g,  Helm charts, Docker Compose packag ...
```
Since we've changed the wording here to no longer indicate the file name is "margo.yaml" I think we need to indicate the application description YAML file must be the only YAML file in the package root. Otherwise, suppliers will have to write extra code to examine all the YAML files to figure out which one is the application description YAML if multiple show up there. While technically it's not a lot of work to do that, it is work that shouldn't need to be done.

Silvanoc replied:
@phil-abb I have moved these kind of additions to a separate branch. Therefore resolving this conversation.

Silvanoc commented:
Before marking the PR as "Ready to Merge" I realized that I was mixing concerns (introduction of LinkML and changes in the Application Package Specification) and that I should not. Unfortunately looking at the details to get the branch ready for merging I forgot about it.

[This](https://github.com/margo/specification/pull/62#discussion_r1945221909) and [this](https://github.com/margo/specification/pull/62#discussion_r1945241958) comments from @phil-abb have remind me of that (thanks!). So I am splitting the PR in two: one with the introduction of LinkML without content changes in the specification itself and another one proposing changes to the specification. @arne-broering @g0zilla I hope it is fine for you. Don't worry, your commits will get simply moved to the 2nd PR.

g0zilla committed:
Add Mermaid support to MkDocs configuration

Enabled Mermaid diagrams in mkdocs.yml by adding the necessary plugin. Provided a working example of a Mermaid diagram to demonstrate its usage and clarify

- Updated .gitignore as needed for related files

Signed-off-by: Andreas Ziller <andreas.ziller@siemens.com>
Acked-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

arne-broering committed:
additional small fixes

Signed-off-by: Arne Broering <arne.broering@siemens.com>
Acked-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc commented:
@arne-broering @g0zilla I have move your additions/modifications to the specification content into [this branch](https://github.com/Silvanoc/margo-specification/tree/various-specification-improvements), though I would consider providing the current two commits in two separate PRs...

Silvanoc commented:
@phil-abb @ajcraig with both your approvals I can simply merge myself on Monday, right?

It is Friday evening in my time zone, so I'm no crazy enough to merge now 🙂

ajcraig commented:
> @phil-abb @ajcraig with both your approvals I can simply merge myself on Monday, right?
> 
> It is Friday evening in my time zone, so I'm no crazy enough to merge now 🙂

Absolutely!
At this stage, pre-draft, I feel comfortable with the author pressing the squash and merge. Appreciate the contributions!
